lib/Firtool/Firtool.cpp

```diff
@@ -0,0 +1,246 @@
+      opt.preserveAggregate, firrtl::PreserveAggregate::None));
+  // Only enable expand whens if lower types is also enabled.
+  auto &modulePM = pm.nest<firrtl::CircuitOp>().nest<firrtl::FModuleOp>();
+  modulePM.addPass(firrtl::createExpandWhensPass());
+  modulePM.addPass(firrtl::createSFCCompatPass());
+
+  pm.nest<firrtl::CircuitOp>().addPass(firrtl::createInlinerPass());
+
+  // Preset the random initialization parameters for each module. The current
+  // implementation assumes it can run at a time where every register is
+  // currently in the final module it will be emitted in, all registers have
+  // been created, and no registers have yet been removed.
+  if (opt.isRandomEnabled(FirtoolOptions::RandomKind::Reg))
+    pm.nest<firrtl::CircuitOp>().nest<firrtl::FModuleOp>().addPass(
+        firrtl::createRandomizeRegisterInitPass());
+
+  if (opt.useOldCheckCombCycles) {
+    if (opt.preserveAggregate == firrtl::PreserveAggregate::None)
+      pm.nest<firrtl::CircuitOp>().addPass(firrtl::createCheckCombCyclesPass());
+    else
+      emitWarning(module.getLoc())
+          << "CheckCombCyclesPass doens't support aggregate "
+             "values yet so it is skipped\n";
+  } else
+    pm.nest<firrtl::CircuitOp>().addPass(firrtl::createCheckCombLoopsPass());
+
+  // If we parsed a FIRRTL file and have optimizations enabled, clean it up.
+  if (!opt.disableOptimization)
+    pm.nest<firrtl::CircuitOp>().nest<firrtl::FModuleOp>().addPass(
+        createSimpleCanonicalizerPass());
+
+  // Run the infer-rw pass, which merges read and write ports of a memory with
+  // mutually exclusive enables.
+  if (!opt.disableOptimization)
+    pm.nest<firrtl::CircuitOp>().nest<firrtl::FModuleOp>().addPass(
+        firrtl::createInferReadWritePass());
+
+  if (opt.replSeqMem)
+    pm.nest<firrtl::CircuitOp>().addPass(firrtl::createLowerMemoryPass());
+
+  pm.nest<firrtl::CircuitOp>().addPass(firrtl::createPrefixModulesPass());
+
+  if (!opt.disableOptimization)
+    pm.nest<firrtl::CircuitOp>().addPass(firrtl::createIMConstPropPass());
+
+  pm.addNestedPass<firrtl::CircuitOp>(firrtl::createAddSeqMemPortsPass());
+
+  pm.nest<firrtl::CircuitOp>().addPass(firrtl::createCreateSiFiveMetadataPass(
+      opt.replSeqMem, opt.replSeqMemCircuit, opt.replSeqMemFile));
+
+  pm.addNestedPass<firrtl::CircuitOp>(firrtl::createExtractInstancesPass());
+
+  // Run passes to resolve Grand Central features.  This should run before
+  // BlackBoxReader because Grand Central needs to inform BlackBoxReader where
+  // certain black boxes should be placed.  Note: all Grand Central Taps related
+  // collateral is resolved entirely by LowerAnnotations.
+  pm.addNestedPass<firrtl::CircuitOp>(
+      firrtl::createGrandCentralPass(opt.grandCentralInstantiateCompanionOnly));
+
+  // Read black box source files into the IR.
+  StringRef blackBoxRoot = opt.blackBoxRootPath.empty()
+                               ? llvm::sys::path::parent_path(inputFilename)
```

Comment:
Bummer the input filename is passed this way, but reworking that is perhaps more of a follow-up sort of task.